vendor/models/profile.py

```diff
@@ -109,4 +109,13 @@ def get_cart_items_count(self):
 
     def get_or_create_address(self, address):
         address, created = self.addresses.get_or_create(name=address.address_1, first_name=address.first_name, last_name=address.last_name, address_1=address.address_1, address_2=address.address_2, locality=address.locality, state=address.state, country=address.country, postal_code=address.postal_code, profile=self)
-        return address, created
\ No newline at end of file
+        return address, created
+
+    def get_customer_products(self):
+        """
+        Get all products that a customer profile has ever owned
+        """
+        return set([ product for receipt in self.receipts.filter(status__gte=PurchaseStatus.COMPLETE) for product in receipt.products.all() ])
```

Comment:
Can this be done with just a query?

Reply:
yes, changed to `CustomerProfile.objects.filter(receipts__products__in=offer_products)`

Reply:
Changed the method to:
```
def has_previously_owned_products(self, products):
        """
        Get all products that the customer has purchased and returns True if it has.
        """
        return bool(self.receipts.filter(Q(products__in=products), Q(status__gte=PurchaseStatus.COMPLETE)))
```

I had been using it in the processor like this:
```
if list(set(subscription.offer.products.all()) & self.invoice.profile.get_customer_products()):
```
but changed it to this:
```
if self.invoice.profile.has_previously_owned_products(subscription.offer.products.all()):
```